### PR TITLE
[Data] Add DataIterator.count() to get dataset/shard size

### LINKED ISSUE #37634: [Data] Get Dataset size from DataIterator
### Description

Currently there’s no public API to get dataset size from DataIterator. However, obtaining the dataset size is critical for many operations in ML training, including  configuring the LR scheduler, setting progress bars, setting maximum training steps. 

Currently there are many workaround to get that number, but none of them are satisfactory solution:
```
train_ds_len = len(list(train_ds.iter_torch_batches(batch_size=1)))

# In Ray AIR transformers_trainer.py
train_ds_len = train_ds._base_dataset.count() // session.get_world_size()
```

Introducing APIs like `DataIterator.count()` can solve the above pain points.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by HirokiNariyoshi**
## Description

Add a public `DataIterator.count()` that returns the number of rows the iterator
will yield, so users (notably Ray Train workers that receive a shard via
`ray.train.get_dataset_shard()`) can size LR schedulers, progress bars, or max
training steps without consuming the iterator.

Previously the only options were exhausting the iterator or calling
`it._base_dataset.count()`.

- `DataIteratorImpl` (from `Dataset.iterator()`) delegates to `Dataset.count()`,
  reading Parquet file metadata where possible (no data scan).
  
- `StreamSplitDataIterator` (from `Dataset.streaming_split()`) returns the
  per-shard count for `equal=True` splits (`total_rows // num_splits`, exact
  since the output splitter drops the remainder). For `equal=False` the
  per-split count is only known at runtime, so it raises `NotImplementedError`
  pointing users to `count()` on the source `Dataset`. It also raises if called
  while the split is being actively iterated.
  
- The base `DataIterator.count()` is a `@PublicAPI` method that raises
  `NotImplementedError` by default, so existing third-party subclasses are
  unaffected.

## Related issues

Closes #37634

## Test plan

- [x] `pytest python/ray/data/tests/test_iterator.py::test_iterator_count -v`
- [x] `pytest python/ray/data/tests/test_iterator.py::test_iterator_count_after_transform -v`
- [x] `pytest python/ray/data/tests/test_streaming_integration.py::test_streaming_split_count_equal -v`
- [x] `pytest python/ray/data/tests/test_streaming_integration.py::test_streaming_split_count_not_equal_raises -v`
- [x] `pytest python/ray/data/tests/test_streaming_integration.py::test_streaming_split_count_during_execution_raises -v`

### CODE REVIEW

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 1]
```diff
@@ -175,6 +176,11 @@ def schema(self) -> Optional["Schema"]:
     def get_context(self) -> DataContext:
         return ray.get(self._coord_actor.get_dataset_context.remote())
 
+    def count(self) -> int:
+        """Implements DataIterator. Raises ``NotImplementedError`` unless the
+        split was created with ``equal=True``."""
+        return ray.get(self._coord_actor.count.remote())
```
[the remark this answers is not in the document]
Comment (HirokiNariyoshi) [reply python/ray/data/_internal/iterator/stream_split_iterator.py:202]: Handled in latest commit [c3e1aec]

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 2]
```diff
@@ -271,6 +277,47 @@ def get_dataset_schema(self):
...
+            )
+
+        with self._dataset_state_lock:
+            if self._current_executor is not None and self._current_executor.is_alive():
+                raise RuntimeError(
+                    "Cannot call count() during active dataset execution. "
+                    "Call count() before or after iterating over the dataset, or "
+                    "call count() directly on the source Dataset object."
+                )
+            total_rows = self._source_dataset_count(split_op)
```
[the remark this answers is not in the document]
Comment (HirokiNariyoshi) [reply python/ray/data/_internal/iterator/stream_split_iterator.py]: Handled in latest commit [c3e1aec]

File: python/ray/data/_internal/iterator/stream_split_iterator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HirokiNariyoshi) [reply python/ray/data/_internal/iterator/stream_split_iterator.py:202]: Updated in latest commit [c3e1aec]

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 3]
```diff
@@ -271,6 +298,58 @@ def get_dataset_schema(self):
...
+                raise RuntimeError(
+                    "Cannot call count() during active dataset execution. "
+                    "Call count() before or after iterating over the dataset, or "
+                    "call count() directly on the source Dataset object."
+                )
+            total_rows = self._source_dataset_count()
+
+        # With ``equal=True``, the output splitter drops the remainder so that
+        # every split yields exactly ``total_rows // num_splits`` rows.
+        return total_rows // self._n
```
[the remark this answers is not in the document]
Comment (HirokiNariyoshi) [reply python/ray/data/_internal/iterator/stream_split_iterator.py:315]: Resolved in latest commit [c059133]

**Comment by HirokiNariyoshi** [comment]
Hi, would appreciate a look when someone has a moment, thanks!

**Comment by HirokiNariyoshi** [comment]
> Hi, would appreciate a look when someone has a moment, thanks!

@owenowenisme , tagging you since you reviewed my last PR, would appreciate a look when you have a moment, thanks!

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 4]
```diff
@@ -175,6 +179,25 @@ def schema(self) -> Optional["Schema"]:
     def get_context(self) -> DataContext:
         return ray.get(self._coord_actor.get_dataset_context.remote())
 
+    def count(self) -> int:
+        """Implements DataIterator.
```
Comment (richardliaw) [inline python/ray/data/_internal/iterator/stream_split_iterator.py]: wrong?

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 5]
```diff
@@ -175,6 +179,25 @@ def schema(self) -> Optional["Schema"]:
...
+
+        Raises ``NotImplementedError`` for ``equal=False`` splits (the per-split
```
Comment (richardliaw) [inline python/ray/data/_internal/iterator/stream_split_iterator.py]: shouldn't raise NotImplemented? probably more like RuntimeError

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 6]
```diff
@@ -48,17 +49,20 @@ def create(
             },
         ).remote(base_dataset, n, locality_hints)
 
-        return [StreamSplitDataIterator(coord_actor, i, n) for i in range(n)]
+        equal = getattr(base_dataset._logical_plan.dag, "equal", False)
```
Comment (richardliaw) [inline python/ray/data/_internal/iterator/stream_split_iterator.py]: let's not use getattr

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 7]
```diff
@@ -271,6 +298,63 @@ def get_dataset_schema(self):
...
+        # count, so an epoch can't start mid-count and defeat the guard. The
+        # count executes only once and is memoized, so the lock is held for a
+        # meaningful duration at most once per coordinator.
+        with self._lock:
+            if self._current_executor is not None and self._current_executor.is_alive():
+                raise RuntimeError(
+                    "Cannot call count() during active dataset execution. "
+                    "Call count() before or after iterating over the dataset, or "
+                    "call count() directly on the source Dataset object."
+                )
```
Comment (richardliaw) [inline python/ray/data/_internal/iterator/stream_split_iterator.py]: feel like you should not raise this in the actor but rather on the iterator side

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 8]
```diff
@@ -271,6 +298,63 @@ def get_dataset_schema(self):
...
+        from ray.data.dataset import Dataset
+
+        # ``_base_dataset`` has a terminal ``StreamingSplit`` operator; its
+        # single input dependency is the source to count.
+        split_op = self._base_dataset._logical_plan.dag
+        source_dag = split_op.input_dependencies[0]
+        source_plan = LogicalPlan(source_dag, self._data_context)
+        source_ds = Dataset._from_parent(self._base_dataset, source_plan)
+        self._source_row_count = source_ds.count()
+        return self._source_row_count
```
Comment (richardliaw) [inline python/ray/data/_internal/iterator/stream_split_iterator.py:339]: why can't you just call self._base_dataset.count()

**Review by justinvyu — CHANGES_REQUESTED** [review]
Instead of adding this `DataIterator.count()` API, can you just call `Dataset.count()` outside of the training loop?

Example:
```python
ds = ray.data.read_parquet(...)

rows_per_worker = ds.count() // num_workers

trainer = TorchTrainer(..., scaling_policy=ray.train.ScalingPolicy(num_workers=num_workers), datasets={"train": ds})
trainer.fit()
```

I'd prefer not adding this to the data iterator API if possible, since `count()` will trigger the full dataset execution if you have any 1:many operators (such as `map_batches`) in your pipeline, since it's impossible to know the number of rows added/removed from your custom UDFs. Then, you need to coordinate a single execution to calculate `count()` across the N training workers.

**Comment by HirokiNariyoshi** [comment]
> Instead of adding this `DataIterator.count()` API, can you just call `Dataset.count()` outside of the training loop?
> 
> Example:
> 
> ```python
> ds = ray.data.read_parquet(...)
> 
> rows_per_worker = ds.count() // num_workers
> 
> trainer = TorchTrainer(..., scaling_policy=ray.train.ScalingPolicy(num_workers=num_workers), datasets={"train": ds})
> trainer.fit()
> ```
> 
> I'd prefer not adding this to the data iterator API if possible, since `count()` will trigger the full dataset execution if you have any 1:many operators (such as `map_batches`) in your pipeline, since it's impossible to know the number of rows added/removed from your custom UDFs. Then, you need to coordinate a single execution to calculate `count()` across the N training workers.

Yes, this doesn't avoid the underlying Dataset.count() execution cost. On the "coordinate a single execution across N workers" point, for streaming_split(..., equal=True) shards, count() routes to the shared SplitCoordinator, which computes the source count once, memoizes it, and returns total_rows // num_splits, so N workers calling it don't trigger N executions.

This formalizes what users are already doing via the private workaround in issue #37634
( train_ds._base_dataset.count() // session.get_world_size() )

Counting on the driver works when you still have the Dataset, but inside train_loop_per_worker you only get the shard via ray.train.get_dataset_shard(), so at the moment the only public option is to pass a precomputed size through train_loop_config. This puts count() on the shard itself.

File: python/ray/data/_internal/iterator/stream_split_iterator.py (hunk 8 above)
Comment (HirokiNariyoshi) [reply python/ray/data/_internal/iterator/stream_split_iterator.py:339]: _base_dataset's plan ends in StreamingSplit, which doesn't propagate num_rows. Calling .count() on it would execute the n-way OutputSplitter and hang or return one split's rows. We count the pre-split source instead and return total // n.

**Comment by richardliaw** [comment]
hm the linked issue doesnt seem to have much traction, why don't we wait for more signal before implementing this?
